cdragontoolbox/hashes.py

```diff
@@ -205,38 +210,83 @@ def substitute_basenames(self):
         """Check all basenames in each subdirectory"""
 
         names = set(os.path.basename(p) for p in self.known.values())
-        dirs = self.directory_list()
+        dirs = list(self.directory_list())
+        unknown = self.unknown # global -> local for increased performance
         logger.info(f"substitute basenames: {len(names)} basenames, {len(dirs)} directories")
         for name in progress_iterator(sorted(names)):
-            self.check_iter(f"{d}/{name}" for d in dirs)
-
-    def _substitute_basename_words(self, words):
-        """Replace all words in known basename"""
+            for d in dirs:
+                h = xxh64(name).intdigest() # inline check() for increased performance
+                if h in unknown:
+                    self._add_known(h, name)
+
+    def _substitute_basename_words(self, paths, words, amount=1):
+        """Substitutes {amount} side by side words in all basenames
+        of all given paths (default=1 for simple 1 word substitution)."""
+
+        regex_part = ""
+        format_part = ""
+        for i in range(1, amount):
+            regex_part += r'([-_][^/_.-]+)'
+            format_part += r'{sep}%%s'
+        format_part = ''.join([r'%s%%s', format_part, r'%s'])
+        re_extract = re.compile(''.join([r'([^/_.-]+)(?=', regex_part, r'[^/]*\.[^/]+$)']))
+        temp_formats = set()
+        for path in paths:
+            for m in re_extract.finditer(path):
+                match = ""
+                for i in range(amount):
+                    match += m.group(i+1)
+                temp_formats.add(format_part % (path[:m.start()], path[m.span()[0]+len(match):]))
 
-        re_extract = re.compile(r'([^/_.-]+)(?=[^/]*\.[^/]+$)')
         formats = set()
-        for path in self.known.values():
-            for m in re_extract.finditer(path):
-                formats.add('%s%%s%s' % (path[:m.start()], path[m.end():]))
+        for format in temp_formats:
+            formats.update(format.replace("{sep}", sep) for sep in "-_")
```

Comment:
This oneline is not that longer: `formats = {fmt.replace("{sep}", sep) for fmt in temp_formats for sep in "-_")}